go/graphd/connect.go

```diff
@@ -0,0 +1,192 @@
+		}
+	}
+}
+
+// isConnected checks if this graphd instance is connected to a graphd database.
+// TODO: Currently only a stub.  True case needs a write check (ie. "status()").
+func (g *graphd) isConnected() bool {
+	if g.conn == nil {
+		return false
+	}
+
+	return true
+}
+
+// Connect to a graphd database.  Connect will attempt to connect to all URLs found in the URLs
+// list associated with this graphd instance, retaining the first successful connection.  On failure,
+// an appropriate error code is returned.  If an acquired connection is already present and valid,
+// Connect returns nil.  Timeout is specified in seconds.  A timeout of 0 is treated as no timeout.
+func (g *graphd) Connect(t int) error {
+	// Set timeout if t > 0, otherwise use the zero value (0s) which signals no timeout.
+	var timeout time.Duration
+	if t > 0 {
+		timeout = time.Duration(t) * time.Second
+	}
+
+	// If URLs list is empty, return error.
+	if len(g.urls) == 0 {
+		errStr := fmt.Sprintf("no URL found in %v", g.urls)
+		g.LogErrf(errStr)
+		return errors.New(errStr)
+	}
+
+	// If already connected, return success.
+	if g.isConnected() {
+		g.LogDebugf("already connected to %v", g.conn.RemoteAddr())
+		return nil
+	}
+
+	// In parallel, for each URL in the list, send a connection request along with the timeout.
+	g.LogDebugf("attempting to connect to %v", g.urls)
+	connCh := initConnCh()
+	numDialers := 0
+	for _, url := range g.urls {
+		go g.dial(url, timeout, connCh)
+		numDialers++
+	}
+
+	// Loop and wait for connections from dialers.  Send an int to signal we're waiting for a connection.
+	// Once we've acquired a valid connection, close our signal channel, preventing other dialers from
+	// sending further connections, and return success.
+	for numDialers > 0 {
+		connCh.awaitingConn <- 1
+		conn := <-connCh.res
+		numDialers--
+
+		// If the connection is invalid, continue listening.
+		if conn == nil {
+			continue
+		}
+
+		// Acquired a valid connection.  Set connection instance.
+		g.LogDebugf("successfully connected to %v", conn.RemoteAddr())
+		g.conn = conn
+
+		// Signal dialers that we've acquired a connection.
+		close(connCh.awaitingConn)
+
+		// Return success.
+		return nil
+	}
+
+	// If no valid connection is acuired, return error
+	errStr := fmt.Sprintf("failed to connect to any URL in %v", g.urls)
+	g.LogErr(errStr)
+	return errors.New(errStr)
+}
+
+// Disconnect attempts to close the existing connection to a graphd database.  On success, nil is
+// returned.  On failure, an error is returned.  Regardless if the connection was properly closed,
+// the connection is zeroed out.
+func (g *graphd) Disconnect() error {
+	// If not connected, return success.
+	if !g.isConnected() {
+		g.LogDebug("no connection present")
+		return nil
+	}
+
+	// Retain address for logs.
+	connectedToAddr := g.conn.RemoteAddr()
+
+	// Try to close the existing connection.
+	err := g.conn.Close()
```

Comment:
move this down, zero out after the error check?
The idiomatic thing in Go makes me look for the err check immediately following the error...this is pedantic, sure, but i'm willing to bet you partly signed up for that.. ;)

Reply:
Yea I know what you're saying, but I did this because I didn't want to repeat `g.conn = nil` in the event that Close() fails and we leak a connection.  However, if a little repetition is less irritating than breaking the error/error check idiom, I'm fine with that.  Thoughts?

Reply:
On second thought, what would you say to something like:

```diff
diff --git a/go/graphd/connect.go b/go/graphd/connect.go
index d3cfecf..f0adec6 100644
--- a/go/graphd/connect.go
+++ b/go/graphd/connect.go
@@ -156,12 +156,14 @@ func (g *graphd) Disconnect() error {
                return nil
        }

+       // Zero out the connection on function exit.
+       defer func() { g.conn = nil }()
+
        // Retain address for logs.
        connectedToAddr := g.conn.RemoteAddr()

        // Try to close the existing connection.
        err := g.conn.Close()
-       // Zero out the connection.
        g.conn = nil
        if err != nil {
                errStr := fmt.Sprintf("failed to close existing connection, resource leak: %v", err)
```

Reply:
ah yea..this works too